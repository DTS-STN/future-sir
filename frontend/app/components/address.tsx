import type { ComponentProps, JSX } from 'react';

import { formatAddress } from '~/utils/string-utils';
import { cn } from '~/utils/tailwind-utils';

type Address = {
  addressLine1?: string;
  addressLine2?: string;
  city?: string;
  postalZipCode?: string;
  provinceState?: string;
  country: string;
};

type AddressProps = ComponentProps<'address'> & {
  address: Address;
  /**
   * The format of the address
   *
   * - `standard`: The standard address format, with the address line, city, province/state, postal/zip code, and country.
   * - `alternative`: An alternative address format, with the address line, city, province/state, postal/zip code, and country on separate lines.
   */
  format?: 'standard' | 'alternative';
};

export function Address({ address, format, className, ...rest }: AddressProps): JSX.Element {
  return (
    <address className={cn('whitespace-pre-wrap not-italic', className)} data-testid="address-id" {...rest}>
      {formatAddress(address, format)}
    </address>
  );
}
