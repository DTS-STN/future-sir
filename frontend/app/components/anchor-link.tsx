import type { ComponentProps, JSX, MouseEvent } from 'react';

/**
 * AnchorLinkProps represents the properties for the AnchorLink component.
 * It extends the ComponentProps<'a'> type, omitting the 'href' property,
 * and adds the required 'anchorElementId' property.
 */
export interface AnchorLinkProps extends OmitStrict<ComponentProps<'a'>, 'href'> {
  anchorElementId: string;
}

/**
 * AnchorLink is a React component used to create anchor links that scroll
 * and focus on the specified target element when clicked.
 *
 * @param props - The properties for the AnchorLink component.
 * @returns React element representing the anchor link.
 */
export function AnchorLink({ anchorElementId, children, onClick, ...restProps }: AnchorLinkProps): JSX.Element {
  /**
   * handleOnSkipLinkClick is the click event handler for the anchor link.
   * It prevents the default anchor link behavior, scrolls to and focuses
   * on the target element specified by 'anchorElementId', and invokes
   * the optional 'onClick' callback.
   */
  function handleOnSkipLinkClick(event: MouseEvent<HTMLAnchorElement>) {
    event.preventDefault();
    scrollAndFocusFromAnchorLink(event.currentTarget.href);
    onClick?.(event);
  }

  return (
    <a href={`#${anchorElementId}`} onClick={handleOnSkipLinkClick} data-testid="anchor-link" {...restProps}>
      {children}
    </a>
  );
}

/**
 * Scrolls and focuses on the element identified by the anchor link's hash.
 *
 * @param href - The anchor link URL.
 */
function scrollAndFocusFromAnchorLink(href: string): void {
  if (URL.canParse(href)) {
    const { hash } = new URL(href);

    if (hash) {
      const targetElement = document.querySelector<HTMLElement>(hash);

      if (targetElement) {
        targetElement.scrollIntoView({ behavior: 'smooth' });
        targetElement.focus();
      }
    }
  }
}
