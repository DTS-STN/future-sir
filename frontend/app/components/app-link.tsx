import type { ComponentProps } from 'react';

import type { Params, Path } from 'react-router';
import { generatePath, Link } from 'react-router';

import { CodedError } from '~/errors/coded-error';
import { ErrorCodes } from '~/errors/error-codes';
import { useLanguage } from '~/hooks/use-language';
import type { I18nRouteFile } from '~/i18n-routes';
import { i18nRoutes } from '~/i18n-routes';
import { getRouteByFile } from '~/utils/route-utils';

type BilingualLinkProps = Omit<ComponentProps<typeof Link>, 'to'> & {
  file: I18nRouteFile;
  hash?: Path['hash'];
  lang?: Language;
  params?: Params;
  search?: Path['search'];
  to?: never;
};

type UnilingualLinkProps = Omit<ComponentProps<typeof Link>, 'lang'> & {
  file?: never;
  hash?: never;
  lang?: Language;
  params?: never;
  search?: never;
};

type AppLinkProps = BilingualLinkProps | UnilingualLinkProps;

export function AppLink({ children, hash, lang, params, file, search, to, ...props }: AppLinkProps) {
  const { currentLanguage } = useLanguage();

  if (to !== undefined) {
    return (
      <Link lang={lang} to={to} {...props}>
        {children}
      </Link>
    );
  }

  const targetLanguage = lang ?? currentLanguage;

  if (targetLanguage === undefined) {
    throw new CodedError(
      'The `lang` parameter was not provided, and the current language could not be determined from the request',
      ErrorCodes.MISSING_LANG_PARAM,
    );
  }

  const route = getRouteByFile(file, i18nRoutes);
  const pathname = generatePath(route.paths[targetLanguage], params);

  const langProp = targetLanguage !== currentLanguage ? targetLanguage : undefined;
  const reloadDocumentProp = props.reloadDocument ?? lang !== undefined;

  return (
    <Link lang={langProp} to={{ hash, pathname, search }} reloadDocument={reloadDocumentProp} {...props}>
      {children}
    </Link>
  );
}
