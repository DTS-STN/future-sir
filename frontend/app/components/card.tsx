import type { HTMLProps, JSX } from 'react';

import type { FlipProp, IconProp } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Slot } from '@radix-ui/react-slot';

import { cn } from '~/utils/tailwind-utils';

type CardProps = HTMLProps<HTMLDivElement> & {
  /**
   * Use the asChild prop to change the default rendered element for the one passed
   * as a child, merging their props and behavior. Useful for rendering links.
   */
  asChild?: boolean;
};

/**
 * Card component renders a card with optional child rendering.
 * It supports various customization options.
 */
export function Card({ className, asChild, ...props }: CardProps): JSX.Element {
  const Comp = asChild ? Slot : 'div';
  return (
    <Comp
      className={cn(
        'group block rounded-xs border bg-white shadow-xs [&:is(a)]:hover:bg-gray-50 [&:is(a)]:hover:shadow-md',
        className,
      )}
      {...props}
    />
  );
}

/**
 * CardHeader component renders the header section of a card.
 */
export function CardHeader({ className, ...props }: HTMLProps<HTMLDivElement>) {
  return <div className={cn('flex flex-col space-y-1.5 p-6', className)} {...props} />;
}

type CardTitleProps = HTMLProps<HTMLHeadingElement> & {
  /**
   * Use the asChild prop to change the default rendered element for the one passed as a child, merging their props and behavior.
   */
  asChild?: boolean;
};

/**
 * CardTitle component renders the title of a card.
 * It supports optional child rendering.
 */
export function CardTitle({ asChild, className, ...props }: CardTitleProps) {
  const Comp = asChild ? Slot : 'h3';
  return (
    <Comp
      className={cn(
        'leading-none font-semibold group-[&:is(a)]:text-slate-700 group-[&:is(a)]:underline group-[&:is(a):focus]:text-blue-700 group-[&:is(a):hover]:text-blue-700',
        className,
      )}
      {...props}
    />
  );
}

/**
 * CardDescription component renders the description of a card.
 */
export function CardDescription({ className, ...props }: HTMLProps<HTMLDivElement>) {
  return <div className={cn('text-sm text-black/60', className)} {...props} />;
}

/**
 * CardContent component renders the content section of a card.
 */
export function CardContent({ className, ...props }: HTMLProps<HTMLDivElement>) {
  return <div className={cn('p-6 pt-0', className)} {...props} />;
}

/**
 * CardFooter component renders the footer section of a card.
 */
export function CardFooter({ className, ...props }: HTMLProps<HTMLDivElement>) {
  return <div className={cn('flex items-center p-6 pt-0', className)} {...props} />;
}

/**
 * CardImage component renders an image in a card.
 */
export function CardImage({ className, ...props }: HTMLProps<HTMLImageElement>) {
  return (
    <img
      alt=""
      className={cn('relative aspect-video h-40 w-full object-cover sm:h-48', className)}
      loading="lazy"
      draggable="false"
      {...props}
    />
  );
}

/**
 * CardTag component renders a tag in a card.
 */
export function CardTag({ className, ...props }: HTMLProps<HTMLSpanElement>) {
  return <span className={cn('inline-block bg-cyan-700 px-2 py-1 text-xs text-white', className)} {...props} />;
}

type CardIconProps = OmitStrict<HTMLProps<HTMLDivElement>, 'children'> & {
  /**
   * FontAwesome icon to be displayed.
   */
  icon: IconProp;
  /**
   * Additional class names for styling the icon.
   */
  iconClassName?: IconProp;
  /**
   * Flip property for the icon.
   */
  iconFlip?: FlipProp;
};

/**
 * CardIcon component renders an icon in a card.
 */
export function CardIcon({ className, icon, iconClassName, iconFlip, ...props }: CardIconProps) {
  return (
    <div className="flex size-10 shrink-0 items-center justify-center rounded-full bg-slate-700 sm:size-12" {...props}>
      <FontAwesomeIcon icon={icon} flip={iconFlip} className={cn('size-4 text-white sm:size-5', iconClassName)} />
    </div>
  );
}
