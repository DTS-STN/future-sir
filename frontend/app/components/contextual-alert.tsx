import type { ReactNode } from 'react';

import {
  faCheckCircle,
  faCircleInfo,
  faCommentDots,
  faExclamationCircle,
  faExclamationTriangle,
} from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { cn } from '~/utils/tailwind-utils';

export type AlertType = 'warning' | 'success' | 'danger' | 'info' | 'comment';

export interface ContextualAlertProps {
  children: ReactNode;
  type: AlertType;
}

const alertBackgroundColors: Partial<Record<AlertType, string>> & { default: string } = {
  comment: 'bg-sky-50',
  default: 'bg-white',
};

const alertBorderColors: Partial<Record<AlertType, string>> & { default: string } = {
  comment: 'border-l-sky-800',
  danger: 'border-l-red-700',
  default: 'border-l-gray-700',
  info: 'border-l-cyan-700',
  warning: 'border-l-amber-700',
  success: 'border-l-green-700',
};

export function ContextualAlert(props: ContextualAlertProps) {
  const { children, type } = props;

  const alertBackgroundColor = alertBackgroundColors[type] ?? alertBackgroundColors.default;
  const alertBorderColor = alertBorderColors[type] ?? alertBorderColors.default;

  return (
    <div className={cn('relative pl-4 sm:pl-6', alertBackgroundColor)}>
      <div className={cn('absolute top-3 left-2 py-1 sm:left-4', alertBackgroundColor)}>
        <Icon type={type} />
      </div>
      <div className={cn('overflow-auto border-l-4 pt-3 pb-2.5 pl-6', alertBorderColor)}>{children}</div>
    </div>
  );
}

function Icon({ type }: { type: string }) {
  switch (type) {
    case 'warning':
      return <FontAwesomeIcon icon={faExclamationTriangle} className="size-5 text-amber-700" data-testid="warning" />;
    case 'success':
      return <FontAwesomeIcon icon={faCheckCircle} className="size-5 text-green-700" data-testid="success" />;
    case 'danger':
      return <FontAwesomeIcon icon={faExclamationCircle} className="size-5 text-red-700" data-testid="danger" />;
    case 'info':
      return <FontAwesomeIcon icon={faCircleInfo} className="size-5 text-cyan-700" data-testid="info" />;
    case 'comment':
      return <FontAwesomeIcon icon={faCommentDots} className="size-5 text-sky-800" data-testid="comment" />;
    default:
      break;
  }
}
