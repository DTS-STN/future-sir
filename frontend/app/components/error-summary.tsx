import { useMemo } from 'react';

import { useTranslation } from 'react-i18next';

type ErrorList = [string?, ...string[]];
type NestedErrors = Readonly<Partial<Record<string, ErrorList | undefined>>>;

interface ErrorSummaryProps {
  errors?: NestedErrors;
  full?: boolean;
}

export function ErrorSummary({ errors, full }: ErrorSummaryProps) {
  const { t } = useTranslation(['gcweb']);

  const summaryErrors = useMemo<ErrorList | undefined>(() => {
    if (errors) {
      return full
        ? Object.values(errors)
            .filter(Array.isArray)
            .filter((arr) => arr.length > 0)
            .flatMap((arr) => arr)
        : Object.values(errors)
            .filter(Array.isArray)
            .filter((arr) => arr.length > 0)
            .map((arr) => arr[0]);
    }
  }, [errors, full]);

  return (
    <>
      {summaryErrors && summaryErrors.length > 0 && (
        <section className="my-5 border-4 border-red-600 p-4" tabIndex={-1} role="alert">
          <h2 className="font-lato text-lg font-semibold">
            {t('gcweb:error-summary.header', { count: summaryErrors.length })}
          </h2>
          <ul className="mt-1.5 list-disc space-y-2 pl-7">
            {summaryErrors.map((errorMessage) => (
              <li key={errorMessage} className="text-red-700 underline">
                {errorMessage}
              </li>
            ))}
          </ul>
        </section>
      )}
    </>
  );
}
