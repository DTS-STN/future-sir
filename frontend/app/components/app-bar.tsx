import { useTranslation } from 'react-i18next';

import { LanguageSwitcher } from '~/components/language-switcher';
import { Menu } from '~/components/menu';
import { UserButton } from '~/components/user-button';

interface AppBarProps {
  children: React.ReactNode;
  name?: string;
  profileItems?: React.ReactNode;
}

export function AppBar({ children, name, profileItems }: AppBarProps) {
  const { t } = useTranslation(['gcweb']);

  return (
    <div className="bg-slate-700">
      <div className="align-center container mx-auto flex flex-wrap justify-between">
        <div className="align-center flex">
          <Menu>{children}</Menu>
        </div>
        <div className="flex items-center space-x-4 text-right">
          {name && <UserButton name={name}>{profileItems}</UserButton>}
          <LanguageSwitcher>{t('gcweb:language-switcher.alt-lang')}</LanguageSwitcher>
        </div>
      </div>
    </div>
  );
}
