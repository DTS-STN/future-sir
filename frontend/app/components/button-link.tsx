import type { ComponentProps, ElementRef } from 'react';
import { forwardRef } from 'react';

import { AppLink } from '~/components/app-link';
import { cn } from '~/utils/tailwind-utils';

const sizes = {
  xs: 'px-3 py-2 text-xs',
  sm: 'px-3 py-2 text-sm',
  base: 'px-5 py-2.5 text-sm',
  lg: 'px-5 py-3 text-base',
  xl: 'px-6 py-3.5 text-base',
} as const;

const variants = {
  alternative:
    'border-gray-200 bg-white text-gray-900 hover:bg-gray-100 hover:text-blue-700 focus:bg-gray-100 focus:text-blue-700',
  default: 'border-gray-300 bg-gray-200 text-slate-700 hover:bg-neutral-300 focus:bg-neutral-300',
  dark: 'border-gray-800 bg-gray-800 text-white hover:bg-gray-900 focus:bg-gray-900',
  green: 'border-green-700 bg-green-700 text-white hover:bg-green-800 focus:bg-green-800',
  primary: 'border-slate-700 bg-slate-700 text-white hover:bg-sky-800 focus:bg-sky-800',
  red: 'border-red-700 bg-red-700 text-white hover:bg-red-800 focus:bg-red-800',
} as const;

type ButtonLinkProps = ComponentProps<typeof AppLink> & {
  disabled?: boolean;
  pill?: boolean;
  size?: keyof typeof sizes;
  variant?: keyof typeof variants;
};

/**
 * Tailwind CSS Buttons from Flowbite
 * @see https://flowbite.com/docs/components/buttons/
 *
 * Disabling a link
 * @see https://www.scottohara.me/blog/2021/05/28/disabled-links.html
 */
export const ButtonLink = forwardRef<HTMLAnchorElement | ElementRef<typeof AppLink>, ButtonLinkProps>(
  ({ children, className, disabled, file, pill, size = 'base', variant = 'default', ...props }, ref) => {
    if (disabled) {
      return (
        <a
          ref={ref}
          className={cn(
            cn(
              'inline-flex items-center justify-center rounded border align-middle font-lato no-underline outline-offset-4',
              sizes[size],
              variants[variant],
              pill && 'rounded-full',
              className,
            ),
            'pointer-events-none cursor-not-allowed opacity-70',
          )}
          role="link"
          aria-disabled="true"
          {...props}
        >
          {children}
        </a>
      );
    }

    return (
      <AppLink
        ref={ref}
        className={cn(
          'inline-flex items-center justify-center rounded border align-middle font-lato no-underline outline-offset-4',
          sizes[size],
          variants[variant],
          pill && 'rounded-full',
          className,
        )}
        file={file}
        {...props}
      >
        {children}
      </AppLink>
    );
  },
);

ButtonLink.displayName = 'ButtonLink';
