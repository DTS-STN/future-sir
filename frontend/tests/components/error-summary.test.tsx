import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';

import { ErrorSummary } from '~/components/error-summary';

describe('ErrorSummary', () => {
  it('should not render an error summary when errors are undefined', () => {
    const { container } = render(<ErrorSummary errors={undefined} />);
    expect(container.innerHTML).toMatchSnapshot('expected html');
  });

  it('should not render an error summary when there are no errors', () => {
    const { container } = render(<ErrorSummary errors={{ Test: [] }} />);
    expect(container.innerHTML).toMatchSnapshot('expected html');
  });

  it('should render an error summary with 1 error', () => {
    const { container } = render(<ErrorSummary errors={{ Test: ['Test error 1'] }} />);
    expect(container.innerHTML).toMatchSnapshot('expected html');
  });

  it('should render an error summary with 2 errors', () => {
    const { container } = render(<ErrorSummary full errors={{ Test: ['Test error 1', 'Test error 2'] }} />);
    expect(container.innerHTML).toMatchSnapshot('expected html');
  });
});
