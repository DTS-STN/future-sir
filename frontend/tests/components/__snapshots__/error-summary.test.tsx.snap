// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorSummary > should not render an error summary when errors are undefined > expected html 1`] = `""`;

exports[`ErrorSummary > should not render an error summary when there are no errors > expected html 1`] = `""`;

exports[`ErrorSummary > should render an error summary with 1 error > expected html 1`] = `"<section class="my-5 border-4 border-red-600 p-4" tabindex="-1" role="alert"><h2 class="font-lato text-lg font-semibold">gcweb:error-summary.header</h2><ul class="mt-1.5 list-disc space-y-2 pl-7"><li class="text-red-700 underline">Test error 1</li></ul></section>"`;

exports[`ErrorSummary > should render an error summary with 2 errors > expected html 1`] = `"<section class="my-5 border-4 border-red-600 p-4" tabindex="-1" role="alert"><h2 class="font-lato text-lg font-semibold">gcweb:error-summary.header</h2><ul class="mt-1.5 list-disc space-y-2 pl-7"><li class="text-red-700 underline">Test error 1</li><li class="text-red-700 underline">Test error 2</li></ul></section>"`;
